src/pcnt.rs

```diff
@@ -0,0 +1,578 @@
+
+/// PCNT channel action on control level
+#[derive(Debug, Copy, Clone, Default)]
+pub enum PcntControlMode {
+    /// Keep current count mode
+    Keep,
+    /// Invert current count mode (increase -> decrease, decrease -> increase)
+    #[default]
+    Reverse,
+    /// Hold current count value
+    Disable,
+}
+
+impl From<PcntControlMode> for pcnt_ctrl_mode_t {
+    fn from(value: PcntControlMode) -> Self {
+        match value {
+            PcntControlMode::Keep => pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_KEEP,
+            PcntControlMode::Reverse => pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_INVERSE,
+            PcntControlMode::Disable => pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_HOLD,
+        }
+    }
+}
+
+#[derive(Debug, EnumSetType)]
+#[enumset(repr = "u32")]
+pub enum PcntEvent {
+    /// PCNT watch point event: threshold1 value event
+    Threshold1 = 2, // pcnt_evt_type_t_PCNT_EVT_THRES_1 = 0x04,
+    /// PCNT watch point event: threshold0 value event
+    Threshold0 = 3, // pcnt_evt_type_t_PCNT_EVT_THRES_0 = 0x08,
+    /// PCNT watch point event: Minimum counter value
+    LowLimit = 4, // pcnt_evt_type_t_PCNT_EVT_L_LIM = 0x10,
+    /// PCNT watch point event: Maximum counter value
+    HighLimit = 5, // pcnt_evt_type_t_PCNT_EVT_H_LIM = 0x20,
+    /// PCNT watch point event: counter value zero event
+    Zero = 6, // pcnt_evt_type_t_PCNT_EVT_ZERO = 0x40,
+}
+
+pub type PcntEventType = enumset::EnumSet<PcntEvent>;
+
+/// Pulse Counter configuration for a single channel
+#[derive(Debug, Copy, Clone, Default)]
+pub struct PcntChannelConfig {
+    /// PCNT low control mode
+    pub lctrl_mode: PcntControlMode,
+    /// PCNT high control mode
+    pub hctrl_mode: PcntControlMode,
+    /// PCNT positive edge count mode
+    pub pos_mode: PcntCountMode,
+    /// PCNT negative edge count mode
+    pub neg_mode: PcntCountMode,
+    /// Maximum counter value
+    pub counter_h_lim: i16,
+    /// Minimum counter value
+    pub counter_l_lim: i16,
+}
+
+impl PcntChannelConfig {
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+#[derive(Debug)]
+pub struct PcntDriver<'d> {
+    unit: pcnt_unit_t,
+    _p: PhantomData<&'d mut ()>,
+}
+
+impl<'d> PcntDriver<'d> {
+    pub fn new<PCNT: Pcnt>(_pcnt: impl Peripheral<P = PCNT> + 'd) -> Result<Self, EspError> {
+        Ok(Self {
+            unit: PCNT::unit(),
+            _p: PhantomData,
+        })
+    }
+
+    /// Configure Pulse Counter chanel
+    ///       @note
+    ///       This function will disable three events: PCNT_EVT_L_LIM, PCNT_EVT_H_LIM, PCNT_EVT_ZERO.
+    /// 
+    /// @param channel Channel to configure
+    /// @param pulse_pin Pulse signal input pin
+    /// @param ctrl_pin Control signal input pin
+    /// @param pconfig Reference of PcntConfig
+    /// 
+    /// @note  Set the signal input to PCNT_PIN_NOT_USED if unused.
+    /// 
+    /// returns
+    /// - ()
+    /// - EspError
+    pub fn channel_config<'a>(
```

Comment:
Sure - as per your comment, this method configures a channel of the unit, but in terms of lifetimes it is still incorrect, and needs to be completely redesigned.

Just one example: the pins you are passing in here are with some lifetime `'a` which is not bound to **anything**. Meaning, you can reuse the pins elsewhere while the channel is also using them, which is wrong and UB.

I see two ways to fix that:
* Implement a top-level `PcntChannelDriver` struct which is perhaps too much effort
* Provide all 4 pins and 2 channel configurations on the `new` constructor of the `PcntDriver`

Reply:
The biggest issue is that the same pins can be used for both channels, a quadrature encoder does this with the pins being reversed for the second channel.

Reply:
Fortunately, that's not an issue at all *IF* you take the approach to provide all the pins for all the two channels at construction time of the `PcntDriver`.

Reply:
so supply 1-4 pins and access them by something like an index?

Reply:
or by reference (not mut)?

Reply:
Supply them using `pin: impl Peripheral<P = impl OutputPin> + 'd` which allows for moth moving the pin peripheral as well as passing a `&mut ref` to the pin.

The fact that a single pin will be shared by two channels of the same PCNT driver **is not a problem** because this is by design so all precatuions are taken in the C driver impl. From the point of view of Rust - because a Channel is not a first-class structure - all is OK as the pin is only owned (or `&mut` used) by a single thing - the particular `PcntDriver`.

As for how exactly you are going to model the API to express which pins are shared between channel 1 and channel 2 and which pins are not - you/we need to think creatively how to model it.
But the important point is - all pins will be owned by the PCNT driver or mutably borrowed into it until the driver is alive.

Reply:
BTW: I hope you understand that `pin: impl Peripheral<P = impl OutputPin> + 'd` does not, and **should not** allow passing pins by non-mut ref (e.g. `&pin`). If that was allowed, you can use one and the same pin in multiple drivers, or multiple instances of the same driver, which is not OK.

Reply:
Why do we need to provide the config **after** the driver has been created? Is this having any real use case?

Reply:
No reason that the user can't change the channel config on the fly, while an encoder implementation does not need to do that you can get pretyy creative withe the PCNT hardware - no need to block this from the user.